# MNIST using Neural Networks

Run the statement to import the TensorFlow library:

In [139]:
import tensorflow as tf

Run the statement to load and prepare the [MNIST dataset](http://yann.lecun.com/exdb/mnist/) and to convert the samples from integers to floating-point numbers:

In [140]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0
x_train = x_train.reshape(x_train.shape[0], 28, 28, 1)
x_test = x_test.reshape(x_test.shape[0], 28, 28, 1)
x_train.shape

(60000, 28, 28, 1)

Build the model with a single layer and relu activation function:

In [141]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Conv2D(28, (3, 3), activation='relu', input_shape=(28, 28, 1)),
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(10, activation = 'relu'),
  tf.keras.layers.Dense(30, activation = 'relu'),
  tf.keras.layers.Dense(10, activation='relu'),
  tf.keras.layers.Dense(20, activation='relu'),
    
  tf.keras.layers.Dense(10)
])
# try to add convolutions and change the no of layers and no of nodes/neurons
# to add a new layer
# tf.keras.layers.Dense(10, activation='relu') 
# to add convolutions 
# tf.keras.layers.Conv2D(28, (3, 3), activation='relu', input_shape=(28, 28)) 
# tf.keras.layers.MaxPooling2D((2, 2))

Choosing our loss function, in this case crossentropy:

In [142]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [143]:
model.compile(optimizer='adam',
              loss=loss_fn,
              metrics=['accuracy'])

Adjusting the model parameters to minimize the loss of our model: 

In [144]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
60000/60000 [==============================] - 13s 216us/sample - loss: 0.3357 - acc: 0.8980
Epoch 2/5
60000/60000 [==============================] - 13s 211us/sample - loss: 0.1148 - acc: 0.9653
Epoch 3/5
60000/60000 [==============================] - 15s 250us/sample - loss: 0.0833 - acc: 0.9747
Epoch 4/5
60000/60000 [==============================] - 27s 454us/sample - loss: 0.0658 - acc: 0.9804
Epoch 5/5
60000/60000 [==============================] - 18s 292us/sample - loss: 0.0525 - acc: 0.9846


Checking the model's performance on the Validation-set.

In [145]:
model.evaluate(x_test,  y_test, verbose=2)

10000/10000 - 1s - loss: 0.0814 - acc: 0.9772


[0.08139963894588872, 0.9772]

Try to improve the accuracy to 97% by changing the number of layers and/or adding convolutions.